rfcs/2021-06-01-7694-improve-monitor-build-performance.md

```diff
@@ -0,0 +1,115 @@
+
+### Manually instrument CI to report build performance
+
+Once we were properly utilizing Datadog on our CI runners, we could then instrument CI builds
+directly to report the build time to Datadog, allowing us to track the build performance over time.
+
+This data would be sparse given how many builds are likely to be triggered on a given day, but would
+form the start of tracking build performance holistically.
+
+## Rationale
+
+Currently, as mentioned above, it can take nearly 40 minutes for a clean release build of Vector.
+Depending on the hardware used, this can be reduced to 25 - 30 minutes, but the time spent is
+untenable even at the 25 - 30 minute range.  This build time not only permeates the local
+development experience, but everything happening in CI, as well.  Any step of the lifecycle of a
+change that involves Vector is unnecessarily slowed down.  The build time of Vector is a massive
+force multiplier.
+
+If we changed nothing about the process, we could still successfully work on Vector, albeit with a
+slower rate of change.  Given the need to occasionally pivot to new features, or bug fixes, or
+whatever is the most important task at hand, rebuilding Vector is actually more common than one
+might expect, and thus every move we make -- whether it be working on a new feature, or fixing a bug
+-- will be slowed down by suboptimal build performance.
+
+## Prior Art
+
+Existing prior art matches the path being proposed here.  There is no one solution to build
+performance, as it is a function of many different settings working in concert.
+
+## Drawbacks
+
+There is a risk that the implementation of this proposal potentially results in occasional
+drawbacks, namely two: `sccache` poisoned cache, and unobservable performance regressions.
+
+As `sccache` caches compiled dependencies, there is the risk that a dependency could be compiled
+such that it is misconfigured, or if it was compiled with bad settings, it would then be reused even
+after the issue was fixed, leading to a confusing development experience.  Practically speaking, it
+is simple to clear the cache, but it would be a new potential line item for things to check when
+debugging an unclear/confusing compilation error, or functional bug.
+
+As well, we currently share the same Cargo profile whether we're benchmarking locally or building a
+customer-facing release of Vector.  There is the risk that performance issues which are not observed
+locally using a build-time-friendly Cargo profile may occur when using the customer-facing-release
+Cargo profile.  This should be rare, given that the profile used for customer-facing releases should
+provide _more_ performance, but the inconsistency alone is enough to potentially obscure comparisons.
```

Comment:
I think we can mitigate this risk by doing CI benchmarks with full release settings. Soak tests not so much, since we'll want debug symbols in place.

Reply:
Yeah, this is a fair point.  I'm going to leave this sentiment intact but I'll update it to encompass the idea that, while we may still execute benchmarks in full trim, we'll have to make sure that local benchmark results are checked against CI.